### PR TITLE
accessibility: Split into separate model and view

### PR DESCRIPTION
**Pull request by chris-steele**
Note: this is part 1 of the exercise of splitting this in model/view - this part is just about handling the immediate split, part 2 will be more of a refactor to have the view respond to changes in the model and any other improvements that can be identified.

Therefore when reviewing please do go ahead and point out any obvious issues or 'quick wins' but maybe save anything more than that until part 2.

### CODE REVIEW

File: js/textInputView.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
+define([
+    'core/js/views/questionView'
+], function(QuestionView) {
+
+    var TextInputView = QuestionView.extend({
+
+ 		events: {
```
Comment (oliverfoster) [inline js/textInputView.js]: indentation

File: js/textInputView.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        },
+
+        hideCorrectAnswer: function() {
+            _.each(this.model.get('_items'), function(item, index) {
+                this.$(".textinput-item-textbox").eq(index).val(item.userAnswer);
+            }, this);
+        },
+
+        onInputChanged:function(e) {
+        	var $input = $(e.target);
```
Comment (oliverfoster) [inline js/textInputView.js]: indentation

File: js/textInputView.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+            this.model.set({
+                _isAtLeastOneCorrectSelection: false,
+                _isCorrect: undefined
+            });
+        },
+
+        showCorrectAnswer: function() {
+            
+            if(this.model.get('_answers'))  {
```
Comment (oliverfoster) [inline js/textInputView.js]: Space

File: js/textInputView.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        // This is important and should give the user feedback on how they answered the question
+        // Normally done through ticks and crosses by adding classes
+        showMarking: function() {
+            if (!this.model.get('_canShowMarking')) return;
+
+            _.each(this.model.get('_items'), function(item, i) {
+                var $item = this.$('.textinput-item').eq(i);
+                $item.removeClass('correct incorrect').addClass(item._isCorrect ? 'correct' : 'incorrect');
+            }, this);
```
Comment (oliverfoster) [inline js/textInputView.js]: native forEach, bind

File: js/textInputView.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                    this.$(".textinput-item-textbox").eq(index).val(item._answers[0]);
+                }, this);
+            }
+            
+        },
+
+        hideCorrectAnswer: function() {
+            _.each(this.model.get('_items'), function(item, index) {
+                this.$(".textinput-item-textbox").eq(index).val(item.userAnswer);
+            }, this);
```
Comment (oliverfoster) [inline js/textInputView.js]: native forEach, bind

File: js/textInputView.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                
+                var correctAnswers = this.model.get('_answers');
+                _.each(this.model.get('_items'), function(item, index) {
+                    this.$(".textinput-item-textbox").eq(index).val(correctAnswers[index][0]);
+                }, this);
+                
+            } else {
+                _.each(this.model.get('_items'), function(item, index) {
+                    this.$(".textinput-item-textbox").eq(index).val(item._answers[0]);
+                }, this);
```
Comment (oliverfoster) [inline js/textInputView.js]: native forEach, bind

File: js/textInputView.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        },
+
+        showCorrectAnswer: function() {
+            
+            if(this.model.get('_answers'))  {
+                
+                var correctAnswers = this.model.get('_answers');
+                _.each(this.model.get('_items'), function(item, index) {
+                    this.$(".textinput-item-textbox").eq(index).val(correctAnswers[index][0]);
+                }, this);
```
Comment (oliverfoster) [inline js/textInputView.js]: native forEach, bind

File: js/textInputView.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        setAllItemsEnabled: function(isEnabled) {
+            _.each(this.model.get('_items'), function(item, index) {
+                var $itemInput = this.$('input').eq(index);
+
+                if (isEnabled) {
+                    $itemInput.prop('disabled', false);
+                } else {
+                    $itemInput.prop('disabled', true);
+                }
+            }, this);
```
Comment (oliverfoster) [inline js/textInputView.js]: native forEach, bind

File: js/textInputModel.js [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            var score = questionWeight * numberOfCorrectAnswers / itemLength;
+
+            this.set('_score', score);
+        },
+
+        resetUserAnswer: function() {
+            _.each(this.get('_items'), function(item) {
+                item["_isCorrect"] = false;
+                item["userAnswer"] = "";
+            }, this);
```
Comment (oliverfoster) [inline js/textInputModel.js]: native forEach, bind

File: js/textInputModel.js [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                }
+                numberOfSpecificAnswers++;
+            }, this);
+        },
+
+        checkAnswerIsCorrect: function(possibleAnswers, userAnswer) {
+            var uAnswer = this.cleanupUserAnswer(userAnswer);
+            var matched = _.filter(possibleAnswers, function(cAnswer){
+                return this.cleanupUserAnswer(cAnswer) == uAnswer;
+            }, this);
```
Comment (oliverfoster) [inline js/textInputModel.js]: native filter, bind

File: js/textInputModel.js [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                    item._isCorrect = true;
+                    item._answerIndex = _.indexOf(item["_answers"], this.cleanupUserAnswer(userAnswer));
+                    this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
+                    this.set('_isAtLeastOneCorrectSelection', true);
+                } else {
+                    item._isCorrect = false;
+                    item._answerIndex = -1;
+                }
+                numberOfSpecificAnswers++;
+            }, this);
```
Comment (oliverfoster) [inline js/textInputModel.js]: native forEach, bind

File: js/textInputModel.js [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        },
+
+        // Marks any items which have answers specific to it
+        // (i.e. item has a _answers array)
+        markSpecificAnswers: function() {
+            var numberOfCorrectAnswers = 0;
+            var numberOfSpecificAnswers = 0;
+            _.each(this.get('_items'), function(item, index) {
+                if(!item._answers) return;
```
Comment (oliverfoster) [inline js/textInputModel.js]: space

File: js/textInputModel.js [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+                        usedAnswerIndexes.push(answerIndex);
+                        item._isCorrect = true;
+                        item._answerIndex = answerIndex + TextInputModel.genericAnswerIndexOffset;
+                        numberOfCorrectAnswers++;
+                        this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
+                        this.set('_isAtLeastOneCorrectSelection', true);
+                    }
+                }, this);
+                if(!item._isCorrect) item._isCorrect = false;
+            }, this);
```
Comment (oliverfoster) [inline js/textInputModel.js]: double native forEach, bind

File: js/textInputModel.js [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+        // Allows the learner to give answers into any input, ignoring the order.
+        // (this excludes any inputs which have their own specific answers).
+        markGenericAnswers: function() {
+            var numberOfCorrectAnswers = 0;
+            var correctAnswers = this.get('_answers').slice();
+            var usedAnswerIndexes = [];
+            _.each(this.get('_items'), function(item, itemIndex) {
+                _.each(correctAnswers, function(answerGroup, answerIndex) {
+                    if(this.checkAnswerIsCorrect(answerGroup, item.userAnswer)) {
```
Comment (oliverfoster) [inline js/textInputModel.js]: return early, space

File: js/textInputModel.js [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+            var userAnswer = new Array( items.length );
+            _.each(items, function(item, index) {
+                userAnswer[ item._index ] = item._answerIndex;
+            });
+            this.set("_userAnswer", userAnswer);
+        },
+
+        isCorrect: function() {
+            if(this.get('_answers')) this.markGenericAnswers();
+            else this.markSpecificAnswers();
```
Comment (oliverfoster) [inline js/textInputModel.js]: Add braces to both blocks

File: js/textInputModel.js [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+        //This preserve the state of the users answers for returning or showing the users answer
+        storeUserAnswer: function() {
+            var items = this.get('_items');
+
+            this.isCorrect();
+
+            var userAnswer = new Array( items.length );
+            _.each(items, function(item, index) {
+                userAnswer[ item._index ] = item._answerIndex;
+            });
```
Comment (oliverfoster) [inline js/textInputModel.js]: native forEach

File: js/textInputModel.js [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+            if (this.get('_isRandom') && this.get('_isEnabled')) {
+                this.set("_items", _.shuffle(this.get("_items")));
+            }
+        },
+
+        // Use to check if the user is allowed to submit the question
+        canSubmit: function() {
+            // can submit if every item has user input
+            var canSubmit = _.every(this.get('_items'), function(item) {
+            	return item.userAnswer != "";
```
Comment (oliverfoster) [inline js/textInputModel.js]: !==

File: js/textInputModel.js [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+            });
+
+            this.setQuestionAsSubmitted();
+            this.markQuestion();
+            this.setScore();
+            this.setupFeedback();
+        },
+
+        setupRandomisation: function() {
+            if (this.get('_isRandom') && this.get('_isEnabled')) {
```
Comment (oliverfoster) [inline js/textInputModel.js]: return early

File: js/textInputModel.js [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+                    item._answerIndex = answerIndex;
+                } else if (answerIndex > -1) {
+                    item.userAnswer = item._answers[answerIndex];
+                    item._answerIndex = answerIndex;
+                } else {
+                    if (item.userAnswer === undefined) item.userAnswer = "******";
+                    item._answerIndex = -1;
+                }
+                if (item.userAnswer instanceof Array) item.userAnswer = item.userAnswer[0];
+            });
```
Comment (oliverfoster) [inline js/textInputModel.js]: native foreach

File: js/textInputModel.js [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        },
+
+        setupQuestionItemIndexes: function() {
+            
+            _.each(this.get('_items'), function(item, index) {
+
+                if (item._index === undefined) item._index = index;
+                if (item._answerIndex === undefined) item._answerIndex = -1;
+
+            });
```
Comment (oliverfoster) [inline js/textInputModel.js]: native foreach

**Comment by moloko** [comment]
sorry for all the updates, went in to change the fw dependency and got a bit carried away... would you gents be kind enough to re-review?

**Comment by moloko** [comment]
@oliverfoster I'll revert 678d415 and we can deal with refactoring it properly in stage 2 of this update

**Pull request merged**